File: Python/SqrtNewtonRaphson.py
def sqrt_newton_raphson(number: float, tolerance: float) -> float:
    approx_low = 1.0
    approx_high = number
    while abs(approx_low - approx_high) >= tolerance:
        approx_low = (approx_low + approx_high) / 2
        approx_high = number / approx_low
    return (approx_low + approx_high) / 2

print(sqrt_newton_raphson(3, 0.0001))

File: CPP/SqrtNewtonRaphson.cpp
#include <iostream>
#include <cmath>

using namespace std;

//O(logn)
double sqrtNewtonRaphson(double number, double tolerance) {
    double lowerApprox = 1;
    double upperApprox = number;
    while (abs(lowerApprox - upperApprox) >= tolerance) {
        lowerApprox = (lowerApprox + upperApprox) / 2;
        upperApprox = number / lowerApprox;
    }
    return (lowerApprox + upperApprox) / 2;
}

int main() {
    cout << sqrtNewtonRaphson(3, 0.0001) << endl;
}